File: server/backend/app/admin/api/v1/invitation.py
"""邀请管理 API"""

from typing import Annotated, Any

from fastapi import APIRouter, Query

from backend.app.userecho.schema.invitation import (
    InvitationCreateReq,
    InvitationSchema,
    InvitationUpdateReq,
    InvitationUsageDetailResp,
)
from backend.app.userecho.service.invitation_service import invitation_service
from backend.common.response.response_schema import response_base
from backend.common.security.jwt import CurrentUser
from backend.database.db import CurrentSession


router = APIRouter(prefix="/invitations", tags=["Admin - 邀请管理"])


@router.post("", summary="创建邀请")
async def create_invitation(
    body: InvitationCreateReq,
    db: CurrentSession,
    current_user: Annotated[dict, CurrentUser],
) -> Any:
    """创建邀请链接"""
    creator_id = current_user.id

    invitation, url, short_url, qr_code_url = await invitation_service.create_invitation(db, creator_id, body)

    # 提交事务
    await db.commit()

    # 构造详细响应
    data = {
        **InvitationSchema.model_validate(invitation).model_dump(),
        "url": url,
        "short_url": short_url,
        "qr_code_url": qr_code_url,
        "remaining_usage": invitation.usage_limit - invitation.used_count,
    }

    return response_base.success(data=data)


@router.get("", summary="邀请列表")
async def get_invitation_list(
    db: CurrentSession,
    _: Annotated[dict, CurrentUser],
    status: str | None = Query(None, description="状态筛选"),
    source: str | None = Query(None, description="来源筛选"),
    campaign: str | None = Query(None, description="活动筛选"),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页数量"),
) -> Any:
    """获取邀请列表（分页）"""
    items, total = await invitation_service.get_invitation_list(db, status, source, campaign, page, size)

    # 构造响应
    data = {
        "items": [InvitationSchema.model_validate(item) for item in items],
        "total": total,
        "page": page,
        "size": size,
    }

    return response_base.success(data=data)


@router.get("/{invitation_id}", summary="邀请详情")
async def get_invitation_detail(
    invitation_id: str,
    db: CurrentSession,
    _: Annotated[dict, CurrentUser],
) -> Any:
    """获取邀请详情"""
    invitation = await invitation_service.get_invitation_detail(db, invitation_id)

    # 获取所有 URL
    url, short_url, qr_code_url = invitation_service.get_invitation_urls(invitation.token)

    data = {
        **InvitationSchema.model_validate(invitation).model_dump(),
        "url": url,
        "short_url": short_url,
        "qr_code_url": qr_code_url,
        "remaining_usage": invitation.usage_limit - invitation.used_count,
    }

    return response_base.success(data=data)


@router.get("/{invitation_id}/usage", summary="邀请使用详情")
async def get_invitation_usage(
    invitation_id: str,
    db: CurrentSession,
    _: Annotated[dict, CurrentUser],
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页数量"),
) -> Any:
    """获取邀请的使用记录"""
    invitation = await invitation_service.get_invitation_detail(db, invitation_id)
    usage_records, total, statistics = await invitation_service.get_invitation_usage_records(
        db, invitation_id, page, size
    )

    # 构造响应（包含用户和租户信息）
    from backend.app.admin.crud.crud_user import user_dao
    from backend.app.userecho.crud.crud_tenant import tenant_dao

    usage_list = []
    for record in usage_records:
        # 获取用户信息
        user = await user_dao.get(db, record.user_id)
        user_info = None
        if user:
            user_info = {
                "id": user.id,
                "email": user.email,
                "nickname": user.nickname,
            }

        # 获取租户信息
        tenant_info = None
        if record.created_tenant_id:
            tenant = await tenant_dao.get(db, record.created_tenant_id)
            if tenant:
                tenant_info = {
                    "id": tenant.id,
                    "name": tenant.name,
                }

        usage_list.append(
            {
                **InvitationUsageDetailResp.model_validate(record).model_dump(),
                "user": user_info,
                "created_tenant": tenant_info,
            }
        )

    data = {
        "invitation": InvitationSchema.model_validate(invitation),
        "usage_records": usage_list,
        "statistics": statistics,
        "total": total,
        "page": page,
        "size": size,
    }

    return response_base.success(data=data)


@router.patch("/{invitation_id}", summary="更新邀请")
async def update_invitation(
    invitation_id: str,
    body: InvitationUpdateReq,
    db: CurrentSession,
    _: Annotated[dict, CurrentUser],
) -> Any:
    """更新邀请信息"""
    invitation = await invitation_service.update_invitation(db, invitation_id, body)
    await db.commit()
    return response_base.success(data=InvitationSchema.model_validate(invitation))


@router.delete("/{invitation_id}", summary="删除邀请")
async def delete_invitation(
    invitation_id: str,
    db: CurrentSession,
    _: Annotated[dict, CurrentUser],
) -> Any:
    """删除邀请（软删除，标记为disabled）"""
    success = await invitation_service.delete_invitation(db, invitation_id)
    await db.commit()
    return response_base.success(data={"success": success})

File: front/apps/web-antd/src/api/system/invitation.ts
/**
 * 邀请管理 API（管理端）
 */

import type { PageResult, Result } from '#/api';

import { requestClient } from '#/api/request';

/**
 * 邀请信息
 */
export interface InvitationInfo {
  id: string;
  token: string;
  usage_limit: number;
  used_count: number;
  expires_at: string;
  plan_code: string;
  trial_days: number;
  source?: string;
  campaign?: string;
  creator_id?: number;
  notes?: string;
  status: string;
  created_at: string;
  updated_at?: string;
}

/**
 * 邀请详情（包含URL）
 */
export interface InvitationDetail extends InvitationInfo {
  url: string;
  short_url?: string;
  qr_code_url: string;
  remaining_usage: number;
}

/**
 * 创建邀请请求
 */
export interface CreateInvitationRequest {
  usage_limit?: number;
  expires_days?: number;
  plan_code?: string;
  trial_days?: number;
  source?: string;
  campaign?: string;
  notes?: string;
}

/**
 * 更新邀请请求
 */
export interface UpdateInvitationRequest {
  status?: string;
  usage_limit?: number;
  expires_at?: string;
  notes?: string;
}

/**
 * 邀请使用记录
 */
export interface InvitationUsageRecord {
  id: string;
  invitation_id: string;
  user_id: number;
  registered_email: string;
  ip_address?: string;
  completed_onboarding: boolean;
  created_tenant_id?: string;
  used_at: string;
  user?: {
    id: number;
    email: string;
    nickname: string;
  };
  created_tenant?: {
    id: string;
    name: string;
  };
}

/**
 * 邀请使用详情响应
 */
export interface InvitationUsageResponse {
  invitation: InvitationInfo;
  usage_records: InvitationUsageRecord[];
  statistics: {
    total_used: number;
    completed_onboarding: number;
    conversion_rate: number;
  };
  total: number;
  page: number;
  size: number;
}

/**
 * 查询参数
 */
export interface InvitationQueryParams {
  status?: string;
  source?: string;
  campaign?: string;
  page?: number;
  size?: number;
}

/**
 * 获取邀请列表
 */
export async function getInvitationList(params?: InvitationQueryParams) {
  return requestClient.get<PageResult<InvitationInfo>>(
    '/api/v1/invitations',
    { params },
  );
}

/**
 * 创建邀请
 */
export async function createInvitation(data: CreateInvitationRequest) {
  return requestClient.post<Result<InvitationDetail>>(
    '/api/v1/invitations',
    data,
  );
}

/**
 * 获取邀请详情
 */
export async function getInvitationDetail(id: string) {
  return requestClient.get<Result<InvitationDetail>>(
    `/api/v1/invitations/${id}`,
  );
}

/**
 * 获取邀请使用详情
 */
export async function getInvitationUsage(id: string, params?: { page?: number; size?: number }) {
  return requestClient.get<Result<InvitationUsageResponse>>(
    `/api/v1/invitations/${id}/usage`,
    { params },
  );
}

/**
 * 更新邀请
 */
export async function updateInvitation(id: string, data: UpdateInvitationRequest) {
  return requestClient.patch<Result<InvitationInfo>>(
    `/api/v1/invitations/${id}`,
    data,
  );
}

/**
 * 删除邀请
 */
export async function deleteInvitation(id: string) {
  return requestClient.delete<Result<{ success: boolean }>>(
    `/api/v1/invitations/${id}`,
  );
}
